### PR TITLE
Add filter options for all crs selection widgets, lots of cleanups

### PR DESCRIPTION
**Pull request by nyalldawson**
This is a partial implementation of supporting crs type filtering in the widgets -- we also should be filtering the lists of recent crs to matching ones. 

I'll do that in a follow up PR, but first there's the basic implementation and a bunch of related general CRS API cleanups which will make this easier...

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmalignrasters.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ QVariantMap QgsAlignRastersAlgorithm::processAlgorithm( const QVariantMap &param
   if ( parameters.value( QStringLiteral( "CRS" ) ).isValid() )
   {
     QgsCoordinateReferenceSystem crs = parameterAsCrs( parameters, QStringLiteral( "CRS" ), context );
-    customCRSWkt = crs.toWkt( QgsCoordinateReferenceSystem::WKT_PREFERRED_GDAL );
+    customCRSWkt = crs.toWkt( Qgis::CrsWktVariant::PreferredGDAL );
```
Comment (nirvn) [inline src/analysis/processing/qgsalgorithmalignrasters.cpp:119]: Should it not be PreferredGdal? E.g., we have QgsGdalUtils, which does not type gdal in all-caps.

File: src/core/qgis.h [hunk 2]
```diff
@@ -1804,6 +1804,44 @@ class CORE_EXPORT Qgis
...
+     * Coordinate reference system WKT formatting variants.
+     *
+     * \note Prior to QGIS 3.36 this was available as QgsCoordinateReferenceSystem::WktVariant
+     *
+     * \since QGIS 3.36
+     */
+    enum class CrsWktVariant SIP_MONKEYPATCH_SCOPEENUM_UNNEST( QgsCoordinateReferenceSystem, WktVariant ) : int
+      {
+      Wkt1GDAL SIP_MONKEYPATCH_COMPAT_NAME( WKT1_GDAL ), //!< WKT1 as traditionally output by GDAL, deriving from OGC 01-009. A notable departure from WKT1_GDAL with respect to OGC 01-009 is that in WKT1_GDAL, the unit of the PRIMEM value is always degrees.
+      Wkt1ESRI SIP_MONKEYPATCH_COMPAT_NAME( WKT1_ESRI ), //!< WKT1 as traditionally output by ESRI software, deriving from OGC 99-049.
```
Comment (nirvn) [inline src/core/qgis.h:1832]: Same question as gdal. Wkt1ESRI or Wkt1Esri ?

File: src/core/proj/qgscoordinatereferencesystem.cpp [hunk 3]
```diff
@@ -3055,98 +3055,22 @@ QStringList QgsCoordinateReferenceSystem::recentProjections()
...
 }
 
 void QgsCoordinateReferenceSystem::clearRecentCoordinateReferenceSystems()
 {
-  QgsSettings settings;
-  settings.remove( QStringLiteral( "UI/recentProjectionsAuthId" ) );
-  settings.remove( QStringLiteral( "UI/recentProjectionsWkt" ) );
-  settings.remove( QStringLiteral( "UI/recentProjectionsProj4" ) );
+
 }
```
Comment (nirvn) [inline src/core/proj/qgscoordinatereferencesystem.cpp:3074]: QgsApplication::coordinateReferenceSystemRegistry()->clearRecent() missing?

File: src/core/qgis.h [hunk 4]
```diff
@@ -1804,6 +1804,44 @@ class CORE_EXPORT Qgis
...
+      Wkt2_2015 SIP_MONKEYPATCH_COMPAT_NAME( WKT2_2015 ), //!< Full WKT2 string, conforming to ISO 19162:2015(E) / OGC 12-063r5 with all possible nodes and new keyword names.
+      Wkt2_2015Simplified SIP_MONKEYPATCH_COMPAT_NAME( WKT2_2015_SIMPLIFIED ), //!< Same as WKT2_2015 with the following exceptions: UNIT keyword used. ID node only on top element. No ORDER element in AXIS element. PRIMEM node omitted if it is Greenwich.  ELLIPSOID.UNIT node omitted if it is UnitOfMeasure::METRE. PARAMETER.UNIT / PRIMEM.UNIT omitted if same as AXIS. AXIS.UNIT omitted and replaced by a common GEODCRS.UNIT if they are all the same on all axis.
+      Wkt2_2018 SIP_MONKEYPATCH_COMPAT_NAME( WKT2_2018 ), //!< Alias for WKT2_2019
+      Wkt2_2018Simplified SIP_MONKEYPATCH_COMPAT_NAME( WKT2_2018_SIMPLIFIED ), //!< Alias for WKT2_2019_SIMPLIFIED
```
Comment (rouault) [inline src/core/qgis.h:1836]: Is there a point in providing 2018 constants ? They were just transitional

Comment (nyalldawson) [reply src/core/qgis.h:1836]: @rouault 

It's just maintaining API compatibility -- the original enum was first introduced in 2018 and had JUST the 2018 constants. Then we added the 2019 aliases to them later, and now we have to drag everything forward. 

That said, I could potentially remove them from c++ and monkey patch them into PyQGIS only. I'll do that...!

File: src/analysis/processing/qgsalgorithmalignrasters.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmalignrasters.cpp:119]: Done

File: src/core/proj/qgscoordinatereferencesystem.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/core/proj/qgscoordinatereferencesystem.cpp:3074]: Thanks , fixed

File: src/core/qgis.h (hunk 4 above)
Comment (nyalldawson) [reply src/core/qgis.h:1836]: Done

**Pull request merged**